src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -450,6 +450,9 @@ private bool IsCandidateSymbol(ISymbol memberSymbol)
                         case SymbolKind.Property:
                             return ((IPropertySymbol)memberSymbol).ExplicitInterfaceImplementations.IsEmpty;
 
+                        case SymbolKind.Event:
+                            return ((IEventSymbol)memberSymbol).ExplicitInterfaceImplementations.IsEmpty;
+
                         default:
                             return true;
```

Comment:
this seems to be one of the bug.  why default to 'true'?

Reply:
Looking at this method, it's the opposite of what i would expect.  Because removing something has such a huge impact, i would have expected a whitelist, instead of a blacklist.  A blacklist means when we don't consider a case, we delete user code.  A whitelist means we only touch the cases we've explicitly stated we think are correct to remove.

Reply:
I generally agree that I'd rather have this diagnostic be under-aggressive than over-aggressive.

Reply:
> A whitelist means we only touch the cases we've explicitly stated

I don't understand the suggestion. Can you give precise details on what such a list would be? Current check states all private member symbols are candidates to be marked as unused, and the switch case deals with exception cases where it is okay to have unreferenced private members.

Reply:
Right now, for example, you have to check for things like MethodKind == StaticConstructor/Destructor to say "don't apply to those guys".  That's a black-list.  You're looking for the things you don't support and you're enumerating those.

Instead, i would only enumerate the cases you *do* support.  That means every switch needs to be 'false' by default (or falls out to the bottom which is 'false), and every case needs to only identify a case you *would* support*.  This is a whitelist.  You enumerate the allowed cases.

Reply:
That would just invert the switch cases, without any real value add:
1. Switch sections for field/property/events will not change.
2. Switch section for method with a switch statement on method kind will be converted to list all method kinds except constructor, destructor, static constructor, and move the default case to that one.

Personally, I don't see a value add as I can't think of any unit test that I can add whose behavior will change will your suggestion. That approach also has a converse drawback of false positives (no reports) getting generated rather then false negatives (incorrect reports), former will go completely under the covers as no one will ever report them.

Reply:
> That would just invert the switch cases, without any real value add:

That is not true.  There are many benefits.  For example:
1. if we miss a case, we don't delete user code.
2. if the compiler adds a new symbol type we don't by default assume it's something we can remove.
3. it clearly shows which cases we've thought about, instead of which cases we may have forgotten.

> That approach also has a converse drawback of false positives (no reports) getting generated rather then false negatives (incorrect reports), 

That would literally be 1000% better :).  'no reports' means code keeps working.  'false' reports means we break user code by deleting things that are used.  This is a destructive feature.  Erring on the side of deleting code is irresponsible behavior on the part of this feature.  

> I don't see a value add as I can't think of any unit test that I can add whose behavior will change will your suggestion. T

That may be true today.  But it may not be true tomorrow.  Also, it depends *entirely* on you being able to enumerate all the cases in your head to validate that we haven't missed anything.  This bug *literally* exists because that strategy didn't work the last time around.

As an example, with the above, i have to ask myself "ok... what are all the different methodkinds i can get**?  is this behavior correct for all of them?".  As opposed to being in the state of going "ok, i can validate it is correct for the enumerated cases, if we miss something we can always add later in a safe manner."

Blacklist vs whitelist switches from the code needing perfect vigilance to not break users, to one where mistakes are non-impactful.  It's a clear and distinct improvement, and is usually one of the first design approaches that are chosen on with destructive systems to prevent problems.

--

** Honestly, i don't even know.  I'd have to go look at all the symbols and kinds to validate things.  That's a lot of work and still isn't correct if/when we change things in the future.  Features should try to be safe by default, and should be clear in intent.  That's no the case here.  And that's one reason why this issue was missed, and why it's hard to feel confidence about this approach going forward.

Reply:
> that's one reason why this issue was missed

That is not true - the issue was missed because I goofed up on not writing a test to cover this case. It wouldn't have changed even if it was written with your proposed approach :-)

I am going to move the discussion on this thread to a separate issue as this is a targeted fix for a scenario that was missed, along with a unit test.

Reply:
https://github.com/dotnet/roslyn/issues/30996

Reply:
> That is not true - the issue was missed because I goofed up on not writing a test to cover this case. It wouldn't have changed even if it was written with your proposed approach :-)

Yes it woudl have.  Events would not have been candidates because they were not contained in the whitelist.  So, by default, anything missed would just be ignored.

Reply:
> So, by default, anything missed would just be ignored

Events were deliberately not added to the switch because I explicitly decided all private unused events should be flagged - I also added large number of tests for events to verify they **are** flagged when unused. The only thing that would have prevented this bug is me having the knowledge that event symbols can have explicit interface implementations, not just methods and properties :-)

Reply:
> I explicitly decided all private unused events should be flagged 

But you did so through *implicit fallthrough*.  Which means the information is only in your head, and is not evident in code.  HTat's my point.  In a whitelist, the code informs the reader about the cases handled.  In a blacklist, the absence of code has to be reinterpretted by the reader as to what the final impact is.  

You've flipped this (as i've mentioned several times) from something i can look at and easily get confidence on, to something that is implicit and which reqiures a lot of analysis to ensure there are no other potential missing issues.  This changes the burden entirely and makes things more difficult.  And, as the default here is to consider something a *candidate* instead of *not* a candidate, we default to the *far worse* "false positive" state than "false negative" state.

Reply:
I have switched the analyzer to be more conservative to unblock the PR and left my views in https://github.com/dotnet/roslyn/issues/30996#issuecomment-436498024.